concordium-node/src/bin/cli.rs

```diff
@@ -485,36 +488,105 @@ async fn maybe_do_out_of_band_catchup(
-                    } else {
-                        info!("Could not complete out of band catch-up due to: {:#}.", e);
-                    }
-                } else {
-                    info!("Completed out of band catch-up from {}.", download_url);
-                }
-                // attempt to properly clean up the downloaded file.
-                if let Err(e) = path.close() {
-                    error!("Could not delete the downloaded file: {}", e);
-                }
-            }
-            Err(e) => {
-                // report error, but continue startup
-                error!("Downloading catch-up files failed: {}.", e)
+        info!("Starting out of band catch-up");
+        let genesis_block_hashes = regenesis_arc.blocks.read().unwrap().clone();
+        if let Err(e) =
+            import_missing_blocks(consensus, &genesis_block_hashes, download_url, data_dir_path)
+                .await
+        {
+            if import_stopped.load(atomic::Ordering::Acquire) {
+                info!("Out of band catchup stopped.");
+            } else {
+                info!("Could not complete out of band catch-up due to: {:#}.", e);
             }
+        } else {
+            info!("Completed out of band catch-up from {}.", download_url)
         }
     }
 }
 
-async fn download_catchup_files(
-    download_url: url::Url,
+// An index entry for a chunk of blocks. Its format must correspond to one
+// produced by `database-exporter`.
+#[derive(serde::Deserialize)]
+struct BlockChunkData {
+    // exported chunk of blocks' filename
+    filename:           String,
+    // genesis block index from which relative heights of blocks in the chunk are counted
+    genesis_index:      usize,
+    // relative height of the oldest block stored in the chunk
+    first_block_height: u64,
+    // relative height of the newest block stored in the chunk
+    last_block_height:  u64,
+}
+
+async fn import_missing_blocks(
+    consensus: &ConsensusContainer,
+    genesis_block_hashes: &[HashBytes],
+    index_url: &url::Url,
     data_dir_path: &Path,
-) -> anyhow::Result<TempPath> {
+) -> anyhow::Result<()> {
+    let current_genesis_index = genesis_block_hashes.len() - 1;
+    let last_finalized_block_height = consensus.get_last_finalized_block_height();
+
+    info!("Current genesis index: {}", current_genesis_index);
+    info!("Local last finalized block height: {}", last_finalized_block_height);
+
+    let comments_stream = reqwest::get(index_url.clone())
+        .await?
+        .bytes_stream()
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e));
+    let mut comments_lines = tokio_util::io::StreamReader::new(comments_stream).lines();
+    while let Some(line) = comments_lines.next_line().await? {
+        if line.starts_with("# genesis hash ") {
+            let index_genesis_block_hash = line.strip_prefix("# genesis hash ").unwrap();
+            let genesis_hash = genesis_block_hashes[0].to_string();
```

Comment:
`genesis_block_hashes` isn't coming from the catchup file. It's coming from the consensus through `regenesis_arc`, I believe. I doubt it can ever be empty, as there has to be at least one genesis block in every chain.